### PR TITLE
Features/miner pool

### PR DESCRIPTION
**Pull request by darren0718**
tvm executor reconstruct and the miner operation optimization

### CODE REVIEW

File: cmd/gtas/cli/chain_op_remote.go [hunk 1]
```diff
@@ -192,55 +190,56 @@ func (ca *RemoteChainOpImpl) BlockByHeight(h uint64) *Result {
...
+		Value:    st,
 		Gas:      gas,
-		Gasprice: gasprice,
-		TxType:   types.TransactionTypeMinerApply,
-		Data:     common.ToHex(data),
+		Gasprice: gasPrice,
+		TxType:   types.TransactionTypeStakeAdd,
+		Data:     data,
 	}
+	fmt.Println("ddd", tx.Data, data)
```
Comment (bird512) [inline cmd/gtas/cli/chain_op_remote.go:236]: what is "ddd" mean here?

File: cmd/gtas/cli/rpc_api_gtas.go [hunk 2]
```diff
@@ -137,21 +141,76 @@ func (api *RpcGtasImpl) GetBlockByHash(hash string) (*Result, error) {
...
-		morts = append(morts, *NewMortGageFromMiner(lightInfo))
+	verifierInfo := core.MinerManagerImpl.GetLatestMiner(address, types.MinerTypeVerify)
+	if verifierInfo != nil {
+		morts = append(morts, NewMortGageFromMiner(verifierInfo))
 	}
-	return successResult(morts)
+	minerDetails.Overview = morts
+	// Get details
+	switch detail {
+	case "":
```
Comment (bird512) [inline cmd/gtas/cli/rpc_api_gtas.go:196]: what's this case use for?

File: core/blockchain_action.go [hunk 3]
```diff
@@ -358,6 +358,14 @@ func (chain *FullBlockChain) addBlockOnChain(source string, b *types.Block) (ret
...
 
 func (chain *FullBlockChain) transitAndCommit(block *types.Block) (ok bool, err error) {
+
+	// Check if all txs exist in the pool to ensure the basic validation is done
+	for _, tx := range block.Transactions {
+		if exist, _ := chain.GetTransactionPool().IsTransactionExisted(tx.Hash); !exist {
+			return false, fmt.Errorf("tx is not in the pool %v", tx.Hash.Hex())
+		}
+	}
+
```
Comment (bird512) [inline core/blockchain_action.go:363]: why need to check all transactions in pool?

File: core/miner_base.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+	prefixMiner               = []byte("minfo")
+	prefixDetail              = []byte("dt")
+	prefixPoolProposal        = []byte("p")
+	prefixPoolVerifier        = []byte("v")
+	keyPoolProposalTotalStake = []byte("totalstake")
+)
+
+const (
+	MinMinerStake             = 500 * common.TAS // minimal token of miner can stake
+	MaxMinerStakeAdjustPeriod = 5000000          // maximal token of miner can stake
```
Comment (bird512) [inline core/miner_base.go:38]: this comment is not suitable

File: network/init.go [hunk 5]
```diff
@@ -159,5 +159,8 @@ func genRandomSeeds(seeds []string) []string {
 }
 
 func GetNetInstance() Network {
+	if netServerInstance == nil {
+		return nil
+	}
```
Comment (bird512) [inline network/init.go:164]: why need this if case?

File: cmd/gtas/cli/chain_op_remote.go (hunk 1 above)
Comment (darren0718) [reply cmd/gtas/cli/chain_op_remote.go:236]: just for test

File: core/blockchain_action.go (hunk 3 above)
Comment (darren0718) [reply core/blockchain_action.go:363]: ensure all the transactions are checked

File: core/tvm_executor.go [hunk 6]
```diff
@@ -28,432 +27,432 @@ import (
...
+		_, logs, err := controller.ExecuteABI(tx.Source, contract, string(tx.Data))
+		ret.logs = logs
 		if err != nil {
-			controller.AccountDB.RevertToSnapshot(snapshot)
+			if err.Code == types.SysABIJSONError {
+				ret.setError(fmt.Errorf(err.Message), types.RSAbiError)
+			} else {
+				ret.setError(fmt.Errorf(err.Message), types.RSTvmError)
+			}
+		} else if canTransfer(ss.accountDB, ss.source, tx.Value.Value()) {
```
Comment (bird512) [inline core/tvm_executor.go:255]: canTransfer() check should place before ExecuteABI()

File: core/rewards_manager.go [hunk 7]
```diff
@@ -83,65 +95,105 @@ func (rm *RewardManager) GetRewardTransactionByBlockHash(blockHash []byte) *type
...
+		return
 	}
 	ids := make([][]byte, 0)
-	for n, _ := reader.Read(addr); n > 0; n, _ = reader.Read(addr) {
-		if n != common.AddressLength {
-			Logger.Debugf("ParseRewardTransaction Addr Size:%d Invalid", n)
-			break
+
+	for _, idx := range targetIdxs {
+		if idx > uint16(len(group.Members)) {
```
Comment (bird512) [inline core/rewards_manager.go:167]: should it be as below?
for index, idx := range targetIdxs {
    if index >= uint16(len(group.Members)) {

File: core/tvm_executor.go [hunk 8]
```diff
@@ -28,432 +27,432 @@ import (
...
+func (op *unSupported) ParseTransaction() error {
+	return fmt.Errorf("unSupported tx type %v", op.typ)
+}
+
+func (op *unSupported) Validate() error {
+	return fmt.Errorf("unSupported tx type %v", op.typ)
+}
+
+func (op *unSupported) Transition() *result {
+	return nil
```
Comment (bird512) [inline core/tvm_executor.go:155]: will throw nil error outside

File: core/tvm_executor.go [hunk 9]
```diff
@@ -28,432 +27,432 @@ import (
...
+	// Including the verifying reward and gas fee share
+	for _, addr := range ss.targets {
+		ss.accountDB.AddBalance(addr, ss.reward)
+	}
+	ret.cumulativeGas = new(big.Int)
 
-	var _type, id, value = MinerManagerImpl.Transaction2MinerParams(transaction)
-	amount := new(big.Int).SetUint64(value)
+	// Add the balance of proposer with pack fee for packing the reward tx
+	ss.accountDB.AddBalance(ss.proposal, ss.packFee)
```
Comment (bird512) [inline core/tvm_executor.go:320]: why not treat the packFee as cumulativeGas so that is can be display in the receipt

Comment (darren0718) [reply core/tvm_executor.go:320]: reward for the block

File: core/tvm_executor.go (hunk 8 above)
Comment (darren0718) [reply core/tvm_executor.go:155]: must not call the method forever

File: middleware/types/miner.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+	totalLen := 2 + pkSize + vrfPkSize
+	if len(bs) != 2 && len(bs) != totalLen {
+		return nil, fmt.Errorf("length error")
+	}
+	version := bs[0]
+	if version != PayloadVersion {
+		return nil, fmt.Errorf("version error %v", version)
+	}
+	pks := &MinerPks{
+		MType: MinerType(bs[1]),
```
Comment (liuhaox1) [inline middleware/types/miner.go:160]: minerType should be checked

File: core/miner_operation.go [hunk 11]
```diff
@@ -0,0 +1,426 @@
...
+			ID:          op.addTarget.Bytes(),
+			Stake:       op.value,
+			ApplyHeight: op.height,
+			Type:        op.minerType,
+			Status:      types.MinerStatusPrepare,
+		}
+	}
+	setPks(targetMiner, op.minerPks)
+	// Check the upper bound of stake
+	if !checkUpperBound(targetMiner, op.height) {
```
Comment (liuhaox1) [inline core/miner_operation.go:142]: need >=

File: core/miner_base.go [hunk 12]
```diff
@@ -0,0 +1,300 @@
...
+	} else if op.opVerifyRole() {
+		key = getPoolKey(prefixPoolVerifier, address)
+
+	}
+	op.minerPool.SetDataSafe(minerPoolAddr, key, []byte{1})
+}
+
+func (op *baseOperation) addProposalTotalStake(addStake uint64) {
+	totalStake := getProposalTotalStake(op.minerPool)
+	op.minerPool.SetDataSafe(minerPoolAddr, keyPoolProposalTotalStake, common.Uint64ToByte(addStake+totalStake))
```
Comment (liuhaox1) [inline core/miner_base.go:245]: need check overflow

File: core/miner_operation.go [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+		}
+	}
+	// Update height
+	detail.Height = op.height
+	if err := op.setDetail(op.addTarget, detailKey, detail); err != nil {
+		return err
+	}
+
+	amount := new(big.Int).SetUint64(op.value)
+	if !canTransfer(op.accountDB, op.addSource, amount) {
```
Comment (liuhaox1) [inline core/miner_operation.go]: need check first

File: core/miner_operation.go [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+	} else {
+		// Accumulate the frozen value
+		frozenDetail.Value += op.value
+	}
+	frozenDetail.Height = op.height
+	// Update the frozen detail of target
+	if err := op.setDetail(op.cancelTarget, frozenDetailKey, frozenDetail); err != nil {
+		return err
+	}
+	return nil
```
Comment (liuhaox1) [inline core/miner_operation.go]: where is add balance?

File: consensus/groupsig/groupsig_test.go [hunk 15]
```diff
@@ -42,9 +42,9 @@ func testIDconvert(t *testing.T) {
 }
 
 //测试用衍生随机数生成私钥，从私钥萃取公钥，以及公钥的序列化
```
Comment (liuhaox1) [inline consensus/groupsig/groupsig_test.go]: use english

File: consensus/groupsig/groupsig_test.go [hunk 16]
```diff
@@ -42,9 +42,9 @@ func testIDconvert(t *testing.T) {
...
-func testPubkey(t *testing.T) {
+func TestPubkey(t *testing.T) {
 	fmt.Printf("\nbegin test pub key...\n")
-	t.Log("testPubkey")
+	t.Log("TestPubkey")
 	r := base.NewRand() //生成随机数
```
Comment (liuhaox1) [inline consensus/groupsig/groupsig_test.go]: use english

File: consensus/groupsig/groupsig_test.go [hunk 17]
```diff
@@ -354,7 +355,7 @@ func test(t *testing.T) {
 	fmt.Printf("sizeof(int) =%v.\n", unsafe.Sizeof(tmp_i))
 	testID(t)          //测试从big.Int生成ID，以及ID的序列化
```
Comment (liuhaox1) [inline consensus/groupsig/groupsig_test.go]: use english

File: core/tvm_executor.go [hunk 18]
```diff
@@ -28,432 +27,432 @@ import (
...
-}
+	ss := newStateTransition(accountDB, tx, bh)
 
-func (executor *TVMExecutor) executeMinerCancelStakeTx(accountdb *account.AccountDB, transaction *types.Transaction, height uint64, castor common.Address, gasUsed *types.BigInt) (success bool, cumulativeGasUsed uint64) {
-	Logger.Debugf("Execute miner cancel pledge tx:%s,source: %v\n", transaction.Hash.Hex(), transaction.Source.Hex())
-	success = false
+	if err = ss.ParseTransaction(); err != nil {
+		Logger.Errorf("state transition parse error:tx %v %v", tx.Hash.Hex(), err)
+		return nil, err
+	}
```
Comment (bird512) [inline core/tvm_executor.go]: should not return if parseTransaction failed. it may have a ddos risk

**Review by liuhaox1 — APPROVED** [review]

**Review by liuhaox1 — APPROVED** [review]

**Pull request merged**